pkg/cmd/roachtest-selector/sfselector/selector.go

```diff
@@ -0,0 +1,298 @@
+		-- count by status
+		count(case when status='SUCCESS' then test_name end) as success_count,
+		count(case when status='FAILURE' then test_name end) as failure_count,
+		-- get the first_run and last_run only if the status is not UNKNOWN. This returns nil for runs that have never run
+		min(case when status!='UNKNOWN' then b.first_run end) as first_run,
+		max(case when status!='UNKNOWN' then b.last_run end) as last_run,
+		sum(duration) as total_duration -- the total duration of the test
+	from DATAMART_PROD.TEAMCITY.TESTS t
+	  -- inner join as explained in the beginning 
+		inner join builds b on
+			b.run_id=t.build_id
+  where test_name is not null -- not interested in tests with no names
+	group  by 1
+)
+select
+	test_name,
+  case when
+    -- mark as selected if
+    failure_count > 0 or -- the test has failed at least once in the past "forPastDays" days
+    first_run > dateadd(DAY, ?, current_date()) or -- recently added test - test has not run for more than "firstRunOn" days
+    last_run < dateadd(DAY, ?, current_date()) or -- the test has been run for last "lastRunOn" days
+    last_run is null -- the test is always ignored till now
+  then 'yes' else 'no' end as selected,
+   -- average duration - this is set to 0 if the test is never run (total_runs=0)
+	case when total_runs > 0 then total_duration/total_runs else 0 end as avg_duration,
+	total_runs,
+from test_stats
+order by selected desc, total_runs`
+)
+
+// supported suites
+var suites = map[string]string{
+	"nightly": "roachtest nightly",
+}
+
+// supported clouds
+var clouds = map[string]struct{}{
+	"gce":   {},
+	"aws":   {},
+	"azure": {},
+}
+
+// SelectTestsReq is the request for SelectTests
+type SelectTestsReq struct {
+	ForPastDays, FirstRunOn, LastRunOn, SelectFromSuccessPct int
+	Cloud, Suite                                             string
+	DryRun                                                   bool
+}
+
+// validate validates the SelectTestsReq
+func (r *SelectTestsReq) validate() error {
+	if _, ok := clouds[r.Cloud]; !ok {
+		keys := make([]string, 0)
+		for c := range clouds {
+			keys = append(keys, c)
+		}
+		return fmt.Errorf("unsupported cloud %s! Supported clouds: %v", r.Cloud, keys)
+	}
+	if _, ok := suites[r.Suite]; !ok {
+		return fmt.Errorf("unsupported suite %s", r.Suite)
+	}
+	return nil
+}
+
+// testDetails contains the results fetched from the query
+type testDetails struct {
+	columnHeaders []string
+	testInfos     [][]string // the fetched data
+}
+
+// SelectTests is responsible for:
+// 1. fetching the data from snowflake using the criteria provided
+// 2. uploading the results in CSV format to cloud storage
+func SelectTests(ctx context.Context, req *SelectTestsReq) error {
+	fmt.Printf("getting test details.\n")
+	r, err := getTestDetails(ctx, req)
+	if err != nil {
+		return err
+	}
+	csvBuffer, err := getCSVInMemory(r)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("uploading %s cloud test details to the cloud...\n", req.Cloud)
```

Comment:
* "to GCS" is probably better than "to the cloud", and also more explicit.

* This could be moved to after `the req.DryRun` check to avoid confusion (since nothing is being uploaded to GCS in that case).

Reply:
Will update the name to "IndentifyAndUpdateTests"